sklearn/cluster/_kmeans.py

```diff
@@ -395,14 +396,24 @@ def _kmeans_single_elkan(X, sample_weight, centers_init, max_iter=300,
 
         centers, centers_new = centers_new, centers
 
-        center_shift_tot = (center_shift**2).sum()
-        if center_shift_tot <= tol:
-            if verbose:
-                print(f"Converged at iteration {i}: center shift "
-                      f"{center_shift_tot} within tolerance {tol}.")
-            break
+        if tol == 0:
+            # When tol = 0 we check that labels did not change because
+            # center_shift might not be exactly 0 due to rounding errors.
+            if np.array_equal(labels, labels_old):
+                strict_convergence = True
+                break
+            labels_old = labels.copy()
+        else:
+            center_shift_tot = (center_shift**2).sum()
+            if center_shift_tot <= tol:
+                if verbose:
+                    print(f"Converged at iteration {i}: center shift "
+                          f"{center_shift_tot} within tolerance {tol}.")
+                strict_convergence = center_shift_tot == 0
```

Comment:
I am always a little concerned with comparing floats with 0. Is this a concern here? To be consistent with master would `center_shift_tot <= 0` be better?

Reply:
it was safe here because sum of squares can't be negative even with rounding errors. But it's irrelevant now, I ordered things a bit differently to address some comments